### PR TITLE
Week 3

### PR DESCRIPTION
**Pull request by Erkan-5Minds**

### CODE REVIEW

File: Week_3/plurality/plurality.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
+#include <cs50.h>
+#include <stdio.h>
+#include <string.h>
+
+// Max number of candidates
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c]: Remove comments.

File: Week_3/plurality/plurality.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        printf("Usage: plurality [candidate ...]\n");
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX)
+    {
+        printf("Maximum number of candidates is %i\n", MAX);
+        return 2;
+    }
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c:33]:
```suggestion
    }
    
```

File: Week_3/plurality/plurality.c [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+bool vote(string name)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+    bool CandidateIsCandidateToVote = strcmp(name, candidates[i].name) == 0;
+    if (CandidateIsCandidateToVote)
+        {
+        candidates[i].votes += 1;
+        return true;  
+        }
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c]: Wrong indent.

File: Week_3/plurality/plurality.c [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    // Display winner of election
+    print_winner();
+}
+
+bool vote(string name)
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+    bool CandidateIsCandidateToVote = strcmp(name, candidates[i].name) == 0;
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c]:
```suggestion
    bool candidateIsCandidateToVote = strcmp(name, candidates[i].name) == 0;
```

File: Week_3/plurality/plurality.c [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    if (CandidateIsCandidateToVote)
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c]:
```suggestion
    if (candidateIsCandidateToVote)
```

File: Week_3/plurality/plurality.c [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+{
+    int max_vote = 0;
+    for (int i = 0; i < candidate_count ; i++)
+    {
+        if (candidates[i].votes > max_vote)
+        {
+            max_vote = candidates[i].votes;            
+        }
+        
+    }
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c:81]:
```suggestion
    }
    
```

File: Week_3/plurality/plurality.c [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    for (int i = 0; i < candidate_count ; i++)
+    {
+       if (candidates[i].votes == max_vote)
+       {
+        printf("%s\n", candidates[i].name);
+       }       
+    }
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c]:
```suggestion
    }
    
```

File: Week_3/runoff/runoff.c [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        printf("Usage: runoff [candidate ...]\n");
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX_CANDIDATES)
+    {
+        printf("Maximum number of candidates is %i\n", MAX_CANDIDATES);
+        return 2;
+    }
```
Comment (DevBastiS) [inline Week_3/runoff/runoff.c:44]:
```suggestion
    }
    
```

File: Week_3/runoff/runoff.c [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+            }
+            break;
+        }
+
+        eliminate(min);
+        for (int i = 0; i < candidate_count; i++)
+        {
+            candidates[i].votes = 0;
+        }
+    }
```
Comment (DevBastiS) [inline Week_3/runoff/runoff.c:108]:
```suggestion
    }
    
```

File: Week_3/runoff/runoff.c [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    return 0;
+}
```
Comment (DevBastiS) [inline Week_3/runoff/runoff.c:111]:
```suggestion
}

```

File: Week_3/runoff/runoff.c [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+{
+     for(int i=0;i<candidate_count;i++) 
+     {
+   if
+      ( strcmp(name,candidates[i].name) == 0)
+   { preferences[voter][rank]=i;
+   
+        return true;
+    }
+     }
```
Comment (DevBastiS) [inline Week_3/runoff/runoff.c]: Wrong indent

File: Week_3/runoff/runoff.c [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+     for(int i=0;i<voter_count;i++) 
+    {
+        for(int j=0;j<candidate_count;j++)
+  {
+      if(candidates[i].eliminated==false)
+   { 
+        candidates[preferences[i][j]].votes++;
+    }
+  }
+    }
```
Comment (DevBastiS) [inline Week_3/runoff/runoff.c]: Wrong indent

File: Week_3/runoff/runoff.c [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    
+    return;
+}
+
+bool print_winner(void)
```
Comment (DevBastiS) [inline Week_3/runoff/runoff.c:144]: Wrond indents all over the place please fix then I will review again.

File: Week_3/tideman/tideman.c [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+        printf("Usage: tideman [candidate ...]\n");
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX)
+    {
+        printf("Maximum number of candidates is %i\n", MAX);
+        return 2;
+    }
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c:43]:
```suggestion
    }
    
```

File: Week_3/tideman/tideman.c [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+bool vote(int rank, string name, int ranks[])
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        if(strcmp(name, candidates[i]) == 0)
+        {
+            ranks[rank] = i;
+            return true;
+        }
+    }
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c:96]:
```suggestion
    }
    
```

File: Week_3/tideman/tideman.c [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+
+void record_preferences(int ranks[])
+{
+    for (int i = 0; i < candidate_count; i++)
+    {
+        for (int j = i + 1; j < candidate_count; j++)
+        {
+            preferences[ranks[i]][ranks[j]] += 1;
+        }
+    }
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c:108]:
```suggestion
    }
    
```

File: Week_3/tideman/tideman.c [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+                pair_count++;
+            }
+            else if (preferences[i][j] < preferences[j][i])
+            {
+                pairs[pair_count].winner = j;
+                pairs[pair_count].loser = i;
+                pair_count++;
+            }
+        }
+    }
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c:131]:
```suggestion
    }
    
```

File: Week_3/tideman/tideman.c [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+        for (int j = i + 1; j < pair_count; j++) {
+            if (preferences[pairs[j].winner][pairs[j].loser] > preferences[pairs[max].winner][pairs[max].loser])
+            {
+                max = j;
+            }
+        }
+        pair temp = pairs[i];
+        pairs[i] = pairs[max];
+        pairs[max] = temp;
+    }
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c:151]:
```suggestion
    }
    
```

File: Week_3/tideman/tideman.c [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+    return;
+}
+
+void sort_pairs(void)
+{
+   for (int i = 0; i < pair_count; i++)
+    {
+        int max = i;
+        for (int j = i + 1; j < pair_count; j++) {
+            if (preferences[pairs[j].winner][pairs[j].loser] > preferences[pairs[max].winner][pairs[max].loser])
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c]: Extract to speaking variable.

File: Week_3/tideman/tideman.c [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+
+void lock_pairs(void)
+{
+   for (int i = 0; i < pair_count; i++)
+    {
+        if (!is_circle(pairs[i].loser, pairs[i].winner))
+        {
+            locked[pairs[i].winner][pairs[i].loser] = true;
+        }
+    }
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c:164]:
```suggestion
    }
    
```

File: Week_3/tideman/tideman.c [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+            }
+
+        }
+
+        if (isLoser) continue;
+        if(!isLoser) 
+        {
+            printf("%s\n", candidates[i]);
+        }
+    }
```
Comment (DevBastiS) [inline Week_3/tideman/tideman.c:190]:
```suggestion
    }
    
```

**Review by DevBastiS — CHANGES_REQUESTED** [review]

File: Week_3/plurality/plurality.c [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+        string name = get_string("Vote: ");
+
+        if (!vote(name))
+        {
+            printf("Invalid vote.\n");
+        }
+    }
+
+    print_winner();
+}
```
Comment (stefanluth) [inline Week_3/plurality/plurality.c]: Missing return

File: Week_3/runoff/runoff.c [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+    for(int i=0;i<candidate_count;i++)
+{
+    string most= candidates[i].name ;
+    if(candidates[i].votes>voter_count/2)
+{ 
+    printf("winner %s\n",most);
+}
+
+}
+    return false;
```
Comment (stefanluth) [inline Week_3/runoff/runoff.c]: This function always returns false.

**Review by stefanluth — CHANGES_REQUESTED** [review]
Fix indents.

File: Week_3/plurality/plurality.c [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+            max_vote = candidates[i].votes;            
+        }
+        
+    }
+
+    for (int i = 0; i < candidate_count ; i++)
+    {
+       if (candidates[i].votes == max_vote)
+       {
+        printf("%s\n", candidates[i].name);
```
Comment (DevBastiS) [inline Week_3/plurality/plurality.c]: Wrong indent

File: Week_3/plurality/plurality.c (hunk 7 above)
Comment (DevBastiS) [reply Week_3/plurality/plurality.c]: Still open

File: Week_3/runoff/runoff.c (hunk 8 above)
Comment (DevBastiS) [reply Week_3/runoff/runoff.c:44]: Still open

File: Week_3/runoff/runoff.c (hunk 9 above)
Comment (DevBastiS) [reply Week_3/runoff/runoff.c:108]: Still open

File: Week_3/runoff/runoff.c (hunk 10 above)
Comment (DevBastiS) [reply Week_3/runoff/runoff.c:111]: Still open

File: Week_3/runoff/runoff.c [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+}
+void eliminate(int minvotes)
+{
+    for(int i = 0; i < candidate_count; i++)
+        if (candidates[i].eliminated==false && candidates[i].votes==minvotes)
+        {
+            candidates[i].eliminated=true;
+        }
+    return; 
+}
```
Comment (DevBastiS) [inline Week_3/runoff/runoff.c]: Alot of wrong indents still here

**Review by DevBastiS — CHANGES_REQUESTED** [review]

File: Week_3/plurality/plurality.c [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+#include <string.h>
+#define MAX 9
+
+typedef struct
+{
+    string name;
+    int votes;
+}
+candidate;
+candidate candidates[MAX];
```
Comment (stefanluth) [inline Week_3/plurality/plurality.c]:
```suggestion
typedef struct
{
    string name;
    int votes;
} candidate;

candidate candidates[MAX];
```

**Review by stefanluth — CHANGES_REQUESTED** [review]
Fix indents :)

**Review by stefanluth — APPROVED** [review]

**Review by DevBastiS — APPROVED** [review]

**Pull request merged**